### PR TITLE
 Embed symbols and source in CompileAndVerify

### PR DESCRIPTION
**Pull request by jcouv**
This changes most `CompileAndVerify` tests to emit embedded PDBs, instead of windows PDBs.
This will allow (1) enabling more tests to run on Linux, and (2) debug such emitted assemblies which is useful on Core (see screenshots below).

Two changes are needed to embed symbols and source:
1. change default emit option to use `DebugInformationFormat.Embedded`
2. include all possible syntax trees from the compilation as embedded text (update: only those with non-empty paths)

~~The latter required relaxing a constraint on `EmbeddedText`, because most syntax trees in our tests don't have a name or file path.
Previously, `EmbeddedText` would throw if given an empty `FilePath`. But the documentation for `EmbeddedText.FilePath` said that indicated that it should be the same as `SyntaxTree.FilePath` which allows empty string.~~

Tagging @tmat @nguerrera for `EmbeddedText` change.

Fixes https://github.com/dotnet/roslyn/issues/31640 (clarify whether `EmbeddedText` should allow empty file path)

source

debug

----
Update: to debug into code produced and executed by `CompileAndVerify`, you must name the source file. Here's an example:
```C#
        [Fact]
        public void Test()
        {
            string source = @"
public class C
{
    public static void Main()
    {
        System.Diagnostics.Debugger.Break();
        System.Diagnostics.Debugger.Launch();
    }
}";
            var c = CreateCompilation(source, options: TestOptions.DebugExe, sourceFileName: "file0"); // <-- here
            CompileAndVerify(c, expectedOutput: "");
        }
```

### CODE REVIEW

**Comment by jaredpar** [comment]
FYI: I retriggered a test run on this PR. I wanted to see what the test time difference would be before / after this change. The existing test run was missing some changes from yesterday that made the comparison invalid.

**Comment by jcouv** [comment]
@tmat @nguerrera I'm relaxing some validation and constraints on `EmbeddedText` so that it accepts an empty `FilePath`. This matches `SyntaxTree.FilePath`. 
Please chime in if you have any concerns.

File: src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs [hunk 1]
```diff
@@ -25,6 +26,42 @@ public class PDBTests : CSharpPDBTestBase
...
         #region General
 
+        [Fact]
+        public void EmbeddedTextWithEmptyFilePath()
+        {
+            var c = CreateCompilationWithMscorlib40AndSystemCore(source: "", options: TestOptions.DebugDll);
+            var embeddedTexts = ImmutableArray.Create(EmbeddedText.FromSource("", SourceText.From("hello", Encoding.UTF8)));
+            var peImage = c.EmitToArray(embeddedTexts: embeddedTexts, options: EmitOptions.Default.WithDebugInformationFormat(DebugInformationFormat.Embedded));
+
+            using (var peReader = new PEReader(peImage))
```
Comment (nguerrera) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs]: We should also check that empty path works for Windows pdbs.

**Comment by nguerrera** [comment]
Who assigned the name file0 in the screenshot? The debugger? I notice that it isn't colorized, and I assume that's because it doesn't have .cs extension?

**Comment by nguerrera** [comment]
This is very cool. Looks good to me, just a couple of questions above.

I can't remember why I required non-empty path. It may just have been misguided defensiveness.

**Comment by jcouv** [comment]
> Who assigned the name file0 in the screenshot?

Sorry, the screenshot is not for the latest implementation. In an interim implementation, I tried giving every syntax tree a name (fileN), but that caused some worse side-effects (since syntax trees are immutable, this would create new syntax trees which would no longer be reference-identical to the originals).

Then I switched to allowing empty file path. I'll grab a screenshot for that too.

**Comment by jcouv** [comment]
I've tried this again today with latest "Int Preview" bits. But the debugging experience didn't display source when I used empty filepaths... I'll follow-up with debugger folks.

**Comment by jcouv** [comment]
After discussion with VS debugger folks, they think that blank file paths in `EmbeddedText` was intentional (the debugger looks for file names first, and then verifies hashes). So I reverted that part of the change.
This PR is still quite useful because it allows more tests to be run on Core and it allows debugging of `CompileAndVerify` execution on Core (which is otherwise hard to do). For such debugging, any source you wish to debug should have a file name specified in `CompileAndVerify`.

@jaredpar This is ready for another look. Thanks

**Comment by nguerrera** [comment]
Is it worth clarifying the comment for file path on EmbeddedText. I believe my "See remarks of SyntaxTree" was meant for the other remarks about how this is not necessarily a literal file path and can be interpreted by resolvers differently and is therefore more opaque, not to suggest that empty is also. The next paragraph after the part that says empty is OK:

https://github.com/dotnet/roslyn/blob/00cbaf02686c03e62a2d127fa1bcb055d0e24cc4/src/Compilers/Core/Portable/Syntax/SyntaxTree.cs#L31-L35

**Comment by jcouv** [comment]
Marking this PR as "personal" again. From discussion with Aleksey, it would be better if we could keep coverage on windows PDBs, instead of shifting most of our coverage from windows PDBs to portable PDBs.
I'll talk to Jared about such double-testing.

File: src/Test/Utilities/Portable/Compilation/CompilationExtensions.cs [hunk 2]
```diff
@@ -46,7 +46,8 @@ public static class CompilationExtensions
                     options = (options ?? EmitOptions.Default).WithDebugInformationFormat(DebugInformationFormat.PortablePdb);
                 }
 
-                pdbStream = new MemoryStream();
+                var discretePdb = (object)options != null && options.DebugInformationFormat != DebugInformationFormat.Embedded;
```
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/CompilationExtensions.cs]: Isn't the check on `options` here redundant with the `if` condition?

File: src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs [hunk 3]
```diff
@@ -202,11 +231,22 @@ EmitOptions emitOptions
             EmitOptions emitOptions
         )
         {
+            if (emitOptions is null)
+            {
+                emitOptions = EmitOptions.Default.WithDebugInformationFormat(DebugInformationFormat.Embedded);
+            }
```
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs]: Worry this isn't the right layer to be specifying emit options. It's pretty low and feels like more of a "just follow orders" layer of the stack. Did you consider putting this in a higher layer like `CompilationVerifier'? That seems to have the same single choke point.

**Pull request merged**